### PR TITLE
Merkletree pkg documentation

### PR DESCRIPTION
**Pull request by c633**
Part of #32 
- Unexport functions and types that are not being used outside the package
- Add comments for exporting functions

### CODE REVIEW

File: merkletree/pad.go [hunk 1]
```diff
@@ -10,7 +10,13 @@ import (
 )
 
 var (
+	// ErrorSTRNotFound indicates that the STR of the searching epoch
```
Comment (masomel) [inline merkletree/pad.go]: I would rephrase this to "the STR of a requested epoch"

File: merkletree/pad.go [hunk 2]
```diff
@@ -10,7 +10,13 @@ import (
...
+	// has been removed from the memory, because the snapshot array has exceeded
+	// its maximum length.
```
Comment (masomel) [inline merkletree/pad.go]: This explanation may not make that much sense to a developer since the snapshot array isn't directly exposed (and there's currently no specification for our PAD design). Maybe say something like "the STR has been evicted from memory, because the maximum number of cached PAD snapshots has been exceeded" is a bit clearer?

File: merkletree/pad.go [hunk 3]
```diff
@@ -10,7 +10,13 @@ import (
...
 	ErrorSTRNotFound = errors.New("[merkletree] STR not found")
+
+	// ErrorNilPolicies is used to panic if the PAD was created
+	// with a nil Policies.
```
Comment (masomel) [inline merkletree/pad.go]: I would change this to "a nil Polices struct". Otherwise, this is grammatically awkward even though it's correct terminology.

File: merkletree/pad.go [hunk 4]
```diff
@@ -78,6 +84,10 @@ func (pad *PAD) updateInternal(policies *Policies, epoch uint64) {
 	}
 }
 
+// Update updates the history hash chain by issuing
+// new signed tree root. It may remove some older STRs from
+// memory if the snapshot array exceeded its maximum capacity.
+// The Policies pointer can be nil if the policies does not change.
```
Comment (masomel) [inline merkletree/pad.go]: I would start with more PAD-related terminology here: E.g. "Update generates a new snapshot of the directory. Specifically, it extends the hash chain by issuing..."

File: merkletree/pad.go [hunk 5]
```diff
@@ -90,15 +100,19 @@ func (pad *PAD) Update(policies *Policies) {
 	pad.updateInternal(policies, pad.latestSTR.Epoch+1)
 }
 
+// Set calculates the private index of the given name and
+// inserts the new binding into the tree.
```
Comment (masomel) [inline merkletree/pad.go]: To be a bit more specific about what's going on here, I would clarify that `Set` creates a new index-to-key binding, and that `Set` inserts it into the Merkle tree underlying the PAD.

File: merkletree/pad.go [hunk 6]
```diff
@@ -90,15 +100,19 @@ func (pad *PAD) Update(policies *Policies) {
...
 func (pad *PAD) Set(name string, value []byte) error {
 	index, _ := pad.computePrivateIndex(name, pad.policies.vrfPrivateKey)
 	return pad.tree.Set(index, name, value)
 }
 
+// Lookup searchs the requested name in the latest version of the tree.
```
Comment (masomel) [inline merkletree/pad.go]: s/searchs/searches, and again, I would refer to the PAD as opposed to the tree.

File: merkletree/pad.go [hunk 7]
```diff
@@ -90,15 +100,19 @@ func (pad *PAD) Update(policies *Policies) {
...
 func (pad *PAD) Lookup(name string) (*AuthenticationPath, error) {
 	return pad.LookupInEpoch(name, pad.latestSTR.Epoch)
 }
 
+// LookupInEpoch searchs the requested name in the tree at requested epoch.
```
Comment (masomel) [inline merkletree/pad.go]: s/searchs/searches

File: merkletree/pad.go [hunk 8]
```diff
@@ -110,17 +124,23 @@ func (pad *PAD) LookupInEpoch(name string, epoch uint64) (*AuthenticationPath, e
...
 	}
 	return pad.snapshots[epoch]
 }
 
+// LatestSTR returns the most recent version of the signed tree root.
 func (pad *PAD) LatestSTR() *SignedTreeRoot {
 	return pad.latestSTR
 }
 
+// TB issues a new promise for the new binding and
```
Comment (masomel) [inline merkletree/pad.go]: It may not be clear to a developer reading this what "the new binding" is, so I would make clear that this binding consists of the given name and value params.

File: merkletree/policy.go [hunk 9]
```diff
@@ -8,6 +8,12 @@ import (
 
 type TimeStamp uint64
 
+// Policies is a summary of the directory's
+// current security policies. This includes the public part
+// of the VRF key used to generate private indices,
+// the cryptographic algorithms in use, as well as
+// the protocol version number.
+// See section 3.3 per paper for an detail example.
```
Comment (masomel) [inline merkletree/policy.go]: s/per paper for an detail/of the CONIKS paper for a detailed

I don't want to assume that a developer reading this necessarily knows which paper we're referring to.

File: merkletree/proof.go [hunk 10]
```diff
@@ -36,6 +42,17 @@ func (n *ProofNode) hash(treeNonce []byte) []byte {
...
+// the prefix path between the corresponding leaf node
+// and the root. This is a proof of inclusion
+// or absence of requested index.
+// The proof type can be determined by
+//  if !bytes.Equal(ap.Leaf.Index, ap.LookupIndex) {
+//	 // Proof of absence
+//	} else {
+//	 // Proof of inclusion
+//  }
+// See section 3.2 per paper for detail.
```
Comment (masomel) [inline merkletree/proof.go]: Same here, I wouldn't assume that the developer knows which paper to look for.

File: merkletree/proof.go [hunk 11]
```diff
@@ -59,7 +76,9 @@ func (ap *AuthenticationPath) authPathHash() []byte {
 	return hash
 }
 
-// Verify should be called after the vrf index is verified successfully
+// Verify verifies the authentication path against
```
Comment (masomel) [inline merkletree/proof.go]: "Verify verifies" sounds awkward. I would be a bit more specific about what this function does: "Verify recomputes the tree's root node from the authentication path, and compares it to the tree hash in the corresponding signed tree root".

File: merkletree/str.go [hunk 12]
```diff
@@ -8,12 +8,12 @@ import (
...
 )
 
-// SignedTreeRoot represents a signed tree root, which is generated
+// SignedTreeRoot represents a signed tree root (STR), which is generated
 // at the beginning of every epoch.
 // Signed tree roots contain the current root node,
 // the current and previous epochs, the hash of the
 // previous STR, and its signature.
-// STR should be final
+// STR should be read-only.
```
Comment (masomel) [inline merkletree/str.go]: This seems to imply that the developer should ensure that the STR is read-only. It might be helpful to be more specific as to why/how the developer can ensure this.

File: merkletree/str.go [hunk 13]
```diff
@@ -56,6 +57,8 @@ func (str *SignedTreeRoot) Serialize() []byte {
 	return strBytes
 }
 
+// VerifyHashChain verifies the consistency of the current hash chain,
+// according to section 3.3 per paper.
```
Comment (masomel) [inline merkletree/str.go]: As with a few other cases, I would be a bit more specific here about what `VerifyHashChain` actually does. Right now you're just stating the purpose of the function.

File: merkletree/pad.go [hunk 14]
```diff
@@ -78,6 +83,11 @@ func (pad *PAD) updateInternal(policies *Policies, epoch uint64) {
 	}
 }
 
+// Update generates a new snapshot of the directory.
+// Specifically, it extends the hash chain by issuing
+// new signed tree root. It may remove some older signed tree roots from
+// memory if the cached PAD snapshots exceeded its maximum capacity.
```
Comment (masomel) [inline merkletree/pad.go]: Nitpicking: s/its/the

File: merkletree/pad.go [hunk 15]
```diff
@@ -90,15 +100,23 @@ func (pad *PAD) Update(policies *Policies) {
 	pad.updateInternal(policies, pad.latestSTR.Epoch+1)
 }
 
+// Set computes the private index of the given name using
```
Comment (masomel) [inline merkletree/pad.go]: Do we want to use the term "key" instead of "name" here?

File: merkletree/pad.go [hunk 16]
```diff
@@ -90,15 +100,23 @@ func (pad *PAD) Update(policies *Policies) {
...
 	index, _ := pad.computePrivateIndex(name, pad.policies.vrfPrivateKey)
 	return pad.tree.Set(index, name, value)
 }
 
+// Lookup searches the requested name in the latest snapshot of the PAD.
 func (pad *PAD) Lookup(name string) (*AuthenticationPath, error) {
 	return pad.LookupInEpoch(name, pad.latestSTR.Epoch)
 }
 
+// LookupInEpoch searches the requested name in the snapshot at requested epoch.
```
Comment (masomel) [inline merkletree/pad.go]: Nitpicking: s/requested/the requested

File: merkletree/pad.go [hunk 17]
```diff
@@ -110,17 +128,24 @@ func (pad *PAD) LookupInEpoch(name string, epoch uint64) (*AuthenticationPath, e
 	return ap, nil
 }
 
+// GetSTR returns the signed tree root of the requested epoch.
+// This signed tree root is read from the caching snapshots of the PAD.
```
Comment (masomel) [inline merkletree/pad.go]: s/caching/cached

File: merkletree/proof.go [hunk 18]
```diff
@@ -7,6 +7,12 @@ import (
 	"github.com/coniks-sys/coniks-go/utils"
 )
 
+// ProofNode can be a leaf node or an empty node,
+// which is included in the returned AuthenticationPath
+// of a given index. The type of that node can be determined
+// by IsEmpty value. It also provides an opening of
```
Comment (masomel) [inline merkletree/proof.go]: Nitpicking: s/by/by the

File: merkletree/proof.go [hunk 19]
```diff
@@ -59,7 +75,9 @@ func (ap *AuthenticationPath) authPathHash() []byte {
 	return hash
 }
 
-// Verify should be called after the vrf index is verified successfully
+// Verify recomputes the tree's root node from the authentication path,
+// and compares it to the tree hash in the corresponding signed tree root.
```
Comment (masomel) [inline merkletree/proof.go]: This sentence could be a bit confusing since this function doesn't use any STRs. Maybe mention the `treeHash` parameter explicitly and mention that it has to come from an STR?

File: merkletree/str.go [hunk 20]
```diff
@@ -56,6 +57,10 @@ func (str *SignedTreeRoot) Serialize() []byte {
 	return strBytes
 }
 
+// VerifyHashChain computes the hash of saved STR,
```
Comment (masomel) [inline merkletree/str.go]: Nitpicking: s/of/of a

File: merkletree/str.go [hunk 21]
```diff
@@ -56,6 +57,10 @@ func (str *SignedTreeRoot) Serialize() []byte {
...
+// and compares it to the hash of previous STR included
```
Comment (masomel) [inline merkletree/str.go]: s/the hash of previous STR/the PreviousSTRHash (I think it's ok to just use the field name)

File: merkletree/tb.go [hunk 22]
```diff
@@ -2,12 +2,20 @@ package merkletree
 
 import "github.com/coniks-sys/coniks-go/crypto/sign"
 
+// TemporaryBinding is a protocol extension which
+// serves two main functions: (1) as a signed promise
+// by the key server to include a newly registered
+// (or updated key) in the next epoch, and (2) to allow immediate
+// use of the key without having to wait for the next epoch.
```
Comment (masomel) [inline merkletree/tb.go]: As I mentioned before, I wouldn't use protocol-related terminology here. I think it's fine to say "PAD" instead of "key server" and "inserted" instead of "registered". Also, the use of "key" is ambiguous here, too.

**Comment by c633** [comment]
> The main changes I suggest are to doc.go, which I think could use more detailed explanations of the data structures. Pretty much everything else are nitpicky edits.

Thanks! Sorry about those mistakes (I always try to avoid them when writing but apparently I didn't do it good enough)  :(

> Also, could you please save the new changes in a separate commit?

Sorry, I will do it.

**Comment by masomel** [comment]
> Sorry about those mistakes (I always try to avoid them when writing but apparently I didn't do it good enough) :(

Don't worry, they really are tiny issues!!

File: merkletree/doc.go [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+Package merkletree implements a Merkle prefix tree and related data structures.
+The Merkle prefix tree is one of the most important components of the CONIKS protocol.
+We implemented this data structure separately as a library
+to help other developers use it in their implementation easily.
+
+Persistent Authenticated Dictionary
+
+This module implements a persistent authenticated dictionary (PAD) where the client
+can query a name-to-key binding at a specific epoch and get authenticated answer.
+The PAD is represented by a hash chain committing to the entire history. This hash chain
```
Comment (c633) [inline merkletree/doc.go]: Thanks for clarifying!

**Comment by c633** [comment]
> The main changes I suggest are to doc.go, which I think could use more detailed explanations of the data structures.

I suppose these explanations should go to https://github.com/coniks-sys/coniks-spec. Instead, I think I could write about `Set`, `Index`, `Sign` are done using the current PAD's underlying data structures (the tree or private keys), whereas `Lookup` is performed on the latest version of the tree. Some more detail about how the tree is rotated using the new VRF key, etc

**Comment by masomel** [comment]
> I suppose these explanations should go to https://github.com/coniks-sys/coniks-spec.

The extra explanations I was suggesting aren't supposed to be a data structure specification, but only enough background for a developer to get a better idea of what the code does and the motivation behind it. The specification would have a much greater level of detail.

> I could write about `Set`, `Index`, `Sign` are done using the current PAD's underlying data structures (the tree or private keys), whereas `Lookup` is performed on the latest version of the tree. 

Those are important details since they tell the developer what is happening under the hood, but I think these explanations would belong with the individual methods that they refer to (much like what @Liamsi has pointed out for the protocol API doc as well).

> Some more detail about how the tree is rotated using the new VRF key, etc

This is an interesting point. This detail is really about `pad.reshuffle()`, right? Since this method isn't exported, a developer just reading the doc won't really know where to look. I'm wondering if this detail belongs with `pad.Update()`, which is where the reshuffling could happen (internally). But I think this is a detail a developer should have since it's not an obvious outcome of `Update`.

File: merkletree/doc.go [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+
+This module implements a persistent authenticated dictionary (PAD) data structure.
+
+Merkle Prefix Tree
+
+This module implements the Merkle prefix tree, which is the data structure
+underlying our PAD implementation. It is a binary tree, and that there are
+two types of leaf nodes: empty leaf node and user leaf node. Each node contains
+its prefix index and its level value. It provides methods to insert, update
+(to update an existing key-value pair) and lookup. The tree is append-only,
```
Comment (arlolra) [inline merkletree/doc.go]: Maybe you mean the PAD is append-only?

File: merkletree/proof.go [hunk 25]
```diff
@@ -7,6 +7,12 @@ import (
 	"github.com/coniks-sys/coniks-go/utils"
 )
 
+// ProofNode can be a leaf node or an empty node,
```
Comment (arlolra) [inline merkletree/proof.go]: s/leaf/user/

File: merkletree/doc.go (hunk 24 above)
Comment (c633) [inline merkletree/doc.go]: I think both PAD and the tree are append-only

File: merkletree/proof.go [hunk 26]
```diff
@@ -36,6 +42,16 @@ func (n *ProofNode) hash(treeNonce []byte) []byte {
 	}
 }
 
+// AuthenticationPath is a pruned tree containing
+// the prefix path between the corresponding leaf node
+// (of type ProofNode) and the root. This is a proof
+// of inclusion or absence of requested index.
+// The proof type can be determined by
```
Comment (arlolra) [inline merkletree/proof.go]: Maybe just say, a proof of inclusion is when the leaf index equals the lookup index?

File: merkletree/doc.go (hunk 24 above)
Comment (arlolra) [inline merkletree/doc.go]: You can update the value in the tree and there's no record of its previous content.  It alone isn't, as I understand it.  The PAD, however, is only updated once an epoch and contains a snapshot of previous values.

File: merkletree/doc.go [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+and it does not support the deletion of values.
```
Comment (arlolra) [inline merkletree/doc.go]: >  it does not support the deletion of values

Isn't that what updating does?

The tree does have the property of not being able to remove user leafs.

File: merkletree/str.go [hunk 28]
```diff
@@ -42,8 +43,8 @@ func NewSTR(key sign.PrivateKey, policies Policies, m *MerkleTree, epoch uint64,
 	return str
 }
 
-// Serialize encodes the STR to a byte array with the following format:
-// [epoch, previous epoch, tree hash, previous STR hash, policies serialization]
+// Serialize serializes the signed tree root into
+// the correct format for signing.
```
Comment (arlolra) [inline merkletree/str.go]: I think it'd be preferable if we could say s/the correct/a specified/ but guess that's alright for now.

File: merkletree/version.go [hunk 29]
```diff
@@ -1,5 +1,7 @@
 package merkletree
 
 const (
-	Version = "0.1"
+	// Version indicates the current version of the library.
+	// This is matched with our release version on Github.
+	Version = "0.1.0"
```
Comment (arlolra) [inline merkletree/version.go]: Hmm, this is included in the STR.  Should that really be the library version or a protocol version?

File: merkletree/doc.go (hunk 27 above)
Comment (c633) [inline merkletree/doc.go]: > The tree does have the property of not being able to remove user leafs. 

Yay, I think it exactly describes what I (and @masomel ) want to say.

File: merkletree/version.go (hunk 29 above)
Comment (c633) [inline merkletree/version.go]: Oh, yep, this should be a protocol version.

**Comment by c633** [comment]
> The extra explanations I was suggesting aren't supposed to be a data structure specification, but only enough background for a developer to get a better idea of what the code does and the motivation behind it. The specification would have a much greater level of detail.

Yup, it makes sense to me. I added it back to doc.go in 1c09c97c6074b71bf4835b1d58e9f27d2d717fd1.

File: merkletree/doc.go [hunk 30]
```diff
@@ -6,18 +6,30 @@ to help other developers use it in their implementation easily.
...
-This module implements the Merkle prefix tree data structure. It provides methods to
-insert, update and lookup. The tree is append-only which means there is no ways to delete
-any node from the tree or change a user leaf node to an empty node.
-All the tree hash operations use the hash algorithm provided by our crypto package
+This module implements the Merkle prefix tree, which is the data structure
+underlying our PAD implementation. It is a binary tree, and that there are
+two types of leaf nodes: empty leaf node and user leaf node. Each node contains
+its prefix index and its level value. It provides methods to insert, update
+(to update an existing key-value pair) and lookup.
+The tree has the property of not being able to remove user leafs.
```
Comment (masomel) [inline merkletree/doc.go]: s/leafs/leaves

File: merkletree/pad.go [hunk 31]
```diff
@@ -84,11 +80,16 @@ func (pad *PAD) updateInternal(policies *Policies, epoch uint64) {
...
-// Update generates a new snapshot of the directory.
+// Update generates a new snapshot of the tree.
+// It should be called in the beginning of each epoch.
 // Specifically, it extends the hash chain by issuing
 // a new signed tree root. It may remove some older signed tree roots from
 // memory if the cached PAD snapshots exceeded the maximum capacity.
-// The Policies pointer can be nil if the policies does not change.
+// policies could be nil if the PAD's policies do not change.
+// If the VRF private key is changed (by passing a new Policies),
+// the underlying tree would be reshuffled. It also means
```
Comment (masomel) [inline merkletree/pad.go]: s/would/will

File: merkletree/version.go (hunk 29 above)
Comment (c633) [inline merkletree/version.go]: @masomel : What do you say?

Comment (masomel) [inline merkletree/version.go]: Hmm, `0.1.0` is definitely the library version, I'm not sure how to number our protocol version. I'd say that the version in the USENIX paper is the second iteration, and after working with you this summer, we've reiterated over it, so I'd say we're on `v3`.

Comment (c633) [inline merkletree/version.go]: Does `3.0` sound good to you? This value is definitely the protocol version, right?

Comment (arlolra) [inline merkletree/version.go]: See #121 where I moved this to the protocol package.

**Comment by arlolra** [comment]
I think you should just call in `0.3` and merge it.  Don't let that be the blocker.

**Comment by c633** [comment]
Merged in 18b43656842c4a8db51f3d41774b7c9abe84f1e7.
